src/datadog/metrics.h

```diff
@@ -0,0 +1,69 @@
+#pragma once
+
+// This component provides an interface, `Metric`, and specific classes for
+// Counter and Gauge metrics. A metric has a name, type, and set of key:value
+// tags associated with it. Metrics can be general to APM or language-specific.
+// General metrics have `common` set to `true`, and language-specific metrics
+// have `common` set to `false`.
+
+#include <atomic>
+#include <string>
+#include <vector>
+
+namespace datadog {
+namespace tracing {
+
+class Metric {
+  // The name of the metric that will be published. A transformation occurs
+  // based on the name and whether it is "common" or "language-specific" when it
+  // is recorded.
+  std::string name_;
+  // The type of the metric. This will currently be count or gauge.
+  std::string type_;
+  // Tags associated with this specific instance of the metric.
+  std::vector<std::string> tags_;
+  // This affects the transformation of the metric name, where it can be a
+  // common telemetry metric, or a language-specific metric that is prefixed
+  // with the language name.
+  bool common_;
+
+ protected:
+  std::atomic<uint64_t> value_ = 0;
+  Metric(std::string name, std::string type, std::vector<std::string> tags,
+         bool common);
+
+ public:
+  // Accessors for name, type, tags, common and capture_and_reset_value are used
+  // when producing the JSON message for reporting metrics.
+  std::string name();
+  std::string type();
+  std::vector<std::string> tags();
+  bool common();
+  uint64_t value();
```

Comment:
I will probably be controversial here but we could get rid of those accessors 👀

Reply:
`value()` hides the atomicity of the underlying member, for better or for worse.

My vote is to keep the trivial methods, but I thought the same as you when I initially read it.